java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/aggregation/MinMaxAggFunc.java

```diff
@@ -18,48 +18,74 @@
 import org.apache.flink.api.common.typeinfo.Types;
+import org.apache.flink.table.runtime.typeutils.ExternalTypeInfo;
 import org.apache.flink.table.types.DataType;
 
-import java.util.TreeMap;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
 
 /** Aggregate function that get the min or max. */
-public class MinMaxAggFunc<T extends Comparable<T>>
-        implements AggFunc<T, T, MinMaxAggFunc.MinMaxAccumulator> {
+public class MinMaxAggFunc<T extends Comparable<T>> implements AggFunc<T, T, HashMap<T, Long>> {
 
     private final DataType inDataType;
-    private final boolean isMin;
+    protected final boolean isMin;
 
     public MinMaxAggFunc(DataType inDataType, boolean isMin) {
         this.inDataType = inDataType;
         this.isMin = isMin;
     }
 
     @Override
-    public void add(MinMaxAccumulator accumulator, T value, long timestamp) {
-        accumulator.values.put(value, accumulator.values.getOrDefault(value, 0L) + 1);
+    public void add(HashMap<T, Long> accumulator, T value, long timestamp) {
+        accumulator.put(value, accumulator.getOrDefault(value, 0L) + 1);
     }
 
     @Override
-    public void retract(MinMaxAccumulator accumulator, T value) {
-        final long newCnt = accumulator.values.get(value) - 1;
+    public void merge(HashMap<T, Long> target, HashMap<T, Long> source) {
+        for (Map.Entry<T, Long> entry : source.entrySet()) {
+            target.put(entry.getKey(), target.getOrDefault(entry.getKey(), 0L) + entry.getValue());
+        }
+    }
+
+    @Override
+    public void retract(HashMap<T, Long> accumulator, T value) {
+        final long newCnt = accumulator.get(value) - 1;
         if (newCnt == 0) {
-            accumulator.values.remove(value);
-            return;
+            accumulator.remove(value);
+        } else {
+            accumulator.put(value, newCnt);
+        }
+    }
+
+    @Override
+    public void retractAccumulator(HashMap<T, Long> target, HashMap<T, Long> source) {
+        for (Map.Entry<T, Long> entry : source.entrySet()) {
+            final long newCnt = target.get(entry.getKey()) - entry.getValue();
+            if (newCnt == 0) {
+                target.remove(entry.getKey());
+                continue;
+            }
+            target.put(entry.getKey(), newCnt);
         }
-        accumulator.values.put(value, newCnt);
     }
 
     @Override
-    public T getResult(MinMaxAccumulator accumulator) {
-        if (accumulator.values.isEmpty()) {
+    public T getResult(HashMap<T, Long> accumulator) {
+        if (accumulator.isEmpty()) {
             return null;
         }
 
+        Comparator<T> comparator;
         if (isMin) {
-            return (T) accumulator.values.firstKey();
+            comparator = Comparator.reverseOrder();
         } else {
-            return (T) accumulator.values.lastKey();
+            comparator = Comparator.naturalOrder();
         }
+
+        return accumulator.keySet().stream()
```

Comment:
It might be more efficient to use a TreeSet and cache the result.

Reply:
We can cache the result to optimize the common case (e.g. min is still the min after adding an element) and add a TODO to use TreeSet in the future.

Reply:
I have reused the existing implementation with a TreeMap. This might make the modification faster and more efficient as well.